### PR TITLE
[fio toup] reportqueue: Do not wake up periodically on empty queue

### PR DESCRIPTION
**Pull request by detsch**
If there are no events to be sent, there is no reason to wake up periodically to try to send pending events. In this case, we should only resume execution once a new event is generated, and the thread is notified.

### CODE REVIEW

File: src/libaktualizr/primary/reportqueue.cc [hunk 1]
```diff
@@ -112,6 +122,9 @@ void ReportQueue::flushQueue() {
       LOG_WARNING << "Failed to post update events: " << response.getStatusStr();
     }
     if (delete_events) {
+      if (report_array.size() < static_cast<size_t>(cur_event_number_limit_)) {
+        no_pending_events_ = true;
```
Comment (mike-sul) [inline src/libaktualizr/primary/reportqueue.cc:126]: I am not sure if this is needed or even correct. I think this condition does NOT mean that there is no any update events in the DB.

Comment (detsch) [reply src/libaktualizr/primary/reportqueue.cc:126]: It is fine to remove this. The only downside is doing an additional iteration before doing the `wait` without timeout.

Comment (mike-sul) [reply src/libaktualizr/primary/reportqueue.cc:126]: Yep, my understanding is that there are update events in the DB in this particular case, so ideally we should enforce a new iteration even without waiting with timeouts?

Comment (mike-sul) [reply src/libaktualizr/primary/reportqueue.cc:126]: I think the change is OK. My concern was related to this case https://github.com/foundriesio/aktualizr/blob/45f29eff34188583b13b65b684dccbcb94013191/src/libaktualizr/primary/reportqueue.cc#L110. I thought it is possible that this condition `if (report_array.size() < static_cast<size_t>(cur_event_number_limit_))` is executed after `cur_event_number_limit_` is divided by two within a single iteration.

**Review by mike-sul — APPROVED** [review]

**Comment by mike-sul** [comment]
@detsch BTW, After this change, it looks like there is no point in "waiting" with timeout. If we know there are update events in the DB why "wait" then at all, just keep sending events. If there are no any events in the DB then this new logic makes the reportqueue wait until new event arrives.

File: src/libaktualizr/primary/reportqueue.cc (hunk 1 above)
Comment (detsch) [reply src/libaktualizr/primary/reportqueue.cc:126]: Yep, it should be fine. In that situation, `delete_events` should be `false`, so we won't hit the affected lines.

**Comment by detsch** [comment]
> @detsch BTW, After this change, it looks like there is no point in "waiting" with timeout. If we know there are update events in the DB why "wait" then at all, just keep sending events. If there are no any events in the DB then this new logic makes the reportqueue wait until new event arrives.

If we are sending big blocks of pending events, it may be safe to wait for a while before going to the next block. This may avoid overwhelming the device gateway with events processing if we have many devices in the same situation.
I would keep the current logic unless there is a requirement for faster delivery of pending events.

**Pull request merged**